### PR TITLE
feat(window): enable floating window for osx, pull os-specific window floating logic to separate function

### PR DESCRIPTION
**Pull request by niajane**
-- please test on linux :D Read some things about installing pyobjc being an issue

how do we feel about the separate file :-)

limitations: displays on every "space" but only on one "screen" (this is probably a good thing?); doesn't display over full screen apps (maybe a bad thing, I wanted this)

### CODE REVIEW

**Review by niajane — COMMENTED** [review]
is this black? :D

File: os_specific_window_manager.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+                NSApplication,
+                NSFloatingWindowLevel,
+                NSWindowCollectionBehaviorCanJoinAllSpaces,
+            )
+
+            ns_app = NSApplication.sharedApplication()
+            ns_window = ns_app.windows()[0]
+            ns_window.setLevel_(NSFloatingWindowLevel)
+            ns_window.setCollectionBehavior_(NSWindowCollectionBehaviorCanJoinAllSpaces)
+        except Exception as e:
```
Comment (niajane) [inline os_specific_window_manager.py:25]: improvement: more specific error handling

File: pyproject.toml [hunk 2]
```diff
@@ -21,6 +21,7 @@ pucoti = "pucoti:app"
 python = "^3.12,<3.13"
 pygame-ce = "2.4.1"
 typer = "^0.12.3"
+pyobjc-framework-Cocoa = "10.3.1"
```
Comment (ddorn) [inline pyproject.toml:24]: I confirm that this doesn't work on linux, failing with `PyObjC requires macOS to build`. You should add the dependency only on macOS, using https://python-poetry.org/docs/dependency-specification/#using-environment-markers and having a try: except ModuleNotFoundError: to gate your logic in os_specific.py

File: os_specific_window_manager.py
Comment (ddorn) [inline os_specific_window_manager.py:1]: It seems good to have a separate file for this. Maybe it should be called `os_specific.py` as more things might go there later.

File: pucoti.py [hunk 3]
```diff
@@ -401,16 +398,7 @@ def place_window(window, x: int, y: int):
...
-        # Thanks gpt4! This moves the window while keeping it on the same display.
-        cmd = (
-            """swaymsg -t get_outputs | jq -r \'.[] | select(.focused) | .rect | "\\(.x + %d) \\(.y + %d)"\' | xargs -I {} swaymsg \'[title="PUCOTI"] floating enable, move absolute position {}\'"""
-            % (x, y)
-        )
-        try:
-            subprocess.check_output(cmd, shell=True)
-        except subprocess.CalledProcessError as e:
-            warnings.warn(f"Failed to move window on sway: {e}")
+    set_window_to_float()
```
Comment (ddorn) [inline pucoti.py:401]: Do you want to do it every time the window moves, or only at startup?

File: os_specific_window_manager.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    elif RUNS_ON_SWAY:
+        # Thanks gpt4! This moves the window while keeping it on the same display.
+        cmd = (
+            """swaymsg -t get_outputs | jq -r \'.[] | select(.focused) | .rect | "\\(.x + %d) \\(.y + %d)"\' | xargs -I {} swaymsg \'[title="PUCOTI"] floating enable, move absolute position {}\'"""
+            % (x, y)
+        )
+        try:
+            subprocess.check_output(cmd, shell=True)
+        except subprocess.CalledProcessError as e:
+            warnings.warn(f"Failed to move window on sway: {e}")
```
Comment (ddorn) [inline os_specific_window_manager.py:36]: This does set the window to floating, but it's more of side effect, as sway needs it to set the position. I think this should be in a separate function called `set_window_pos`

Comment (ddorn) [reply os_specific_window_manager.py:36]: Also, it needs access to `x` and `y`

**Review by ddorn — COMMENTED** [review]
Nice addition! I think other mac users wanted similar things.
Also, from reading you, it seems that it's more "enable _sticky_ window* but the terminology might be different on mac :shrug:
It seems fine to not display on every screen (if you mean physical displays) and over fullscreen apps. I you could potentially make this configurable, but it seems a sane default behavior (aka, the one I use).

**Comment by ddorn** [comment]
> is this black? :D

Yes, it is! You can see by clicking the small red cross ❌ what checks failed. You can also have it run automatically on your local machine by running `pre-commit install`, which is part of the dev environment of pucoti. It's pretty nice :D

**Pull request merged**